roc_jni/src/main/cpp/logger.cpp

```diff
@@ -41,85 +40,94 @@ static const char* logLevelMapping(roc_log_level level) {
+static void logger_handler(const roc_log_message* message, void* argument) {
     JNIEnv* env = NULL;
-    jfieldID field = NULL;
-    jobject levelObj = NULL;
-    jstring jmodule = NULL;
-    jstring jmess = NULL;
-
-    handler_args.mutex.lock();
-    if (handler_args.vm == NULL || handler_args.callback == NULL || handler_args.methID == NULL) {
-        handler_args.mutex.unlock();
-        return;
+    jclass levelClass = NULL;
+    jfieldID levelFieldID = NULL;
+    jobject msgLevel = NULL;
+    jstring msgModule = NULL;
+    jstring msgText = NULL;
+    jint result = 0;
+    int detach = 0;
+
+    pthread_mutex_lock(&handler_mutex);
+
+    if (handler.vm == NULL || handler.object == NULL || handler.method == NULL) {
+        goto out;
     }
 
-    jint res;
-    int attached = 0; // know if detaching at the end is necessary
     // check if it is needed to attach current thread
-    if ((res = handler_args.vm->GetEnv((void**) &env, JNI_VERSION)) == JNI_EDETACHED) {
+    if ((result = handler.vm->GetEnv((void**) &env, JNI_VERSION)) == JNI_EDETACHED) {
 #ifdef __ANDROID__
-        if (handler_args.vm->AttachCurrentThread(&env, 0) == JNI_OK)
+        if (handler.vm->AttachCurrentThread(&env, 0) == JNI_OK)
 #else
-        if (handler_args.vm->AttachCurrentThread((void**) &env, 0) == JNI_OK)
+        if (handler.vm->AttachCurrentThread((void**) &env, 0) == JNI_OK)
 #endif
-            attached = 1;
+            detach = 1;
         else {
             // cannot attach current thread
-            handler_args.mutex.unlock();
-            return;
+            goto out;
         }
-    } else if (res != JNI_OK) {
+    } else if (result != JNI_OK) {
         // cannot get env
-        handler_args.mutex.unlock();
-        return;
+        goto out;
     }
 
-    field = env->GetStaticFieldID(
-        handler_args.logLevelClass, logLevelMapping(message->level), "L" LOG_LEVEL_CLASS ";");
-    levelObj = env->GetStaticObjectField(handler_args.logLevelClass, field);
+    assert(env != NULL);
 
-    jmodule = env->NewStringUTF(message->module);
-    jmess = env->NewStringUTF(message->text);
+    levelClass = env->FindClass(LOG_LEVEL_CLASS);
+    assert(levelClass != NULL);
 
-    env->CallVoidMethod(handler_args.callback, handler_args.methID, levelObj, jmodule, jmess);
+    levelFieldID = env->GetStaticFieldID(
+        levelClass, msgLevelMapping(message->level), "L" LOG_LEVEL_CLASS ";");
+    assert(levelFieldID != NULL);
 
-    if (attached) handler_args.vm->DetachCurrentThread();
+    msgLevel = env->GetStaticObjectField(levelClass, levelFieldID);
+    assert(msgLevel != NULL);
 
-    handler_args.mutex.unlock();
+    msgModule = env->NewStringUTF(message->module);
+    assert(msgModule != NULL);
+
+    msgText = env->NewStringUTF(message->text);
+    assert(msgText != NULL);
+
+    env->CallVoidMethod(handler.object, handler.method, msgLevel, msgModule, msgText);
+
+    if (env->ExceptionCheck()) {
+        env->ExceptionDescribe();
+    }
```

Comment:
I think we need to add `env->ExceptionClear();`

```java
    if (env->ExceptionCheck()) {
        env->ExceptionDescribe();
        env->ExceptionClear();
    }
    ```
otherwise exception will be thrown on the next native call

Reply:
Fixed